contracts/test/integration/utils/MintRedeemVerification.sol

```diff
@@ -0,0 +1,157 @@
+pragma solidity =0.8.13;
+
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {MainnetAddresses} from "../fixtures/MainnetAddresses.sol";
+import {ArbitrumAddresses} from "../fixtures/ArbitrumAddresses.sol";
+import {PriceBoundPSM} from "../../../peg/PriceBoundPSM.sol";
+import {Vm} from "./../../unit/utils/Vm.sol";
+import {IVolt} from "../../../volt/Volt.sol";
+import {ICore} from "../../../core/ICore.sol";
+
+import "hardhat/console.sol";
+
+/// @notice contract to verify that all PSM's are able to mint and redeem
+/// after a proposal
+contract MintRedeemVerification {
+    /// TODO add support for arbitrum
+
+    /// @notice all PSM's on mainnet
+    address[] private allMainnetPSMs = [
+        MainnetAddresses.VOLT_DAI_PSM,
+        MainnetAddresses.VOLT_USDC_PSM
+    ];
+
+    /// @notice amount of volt to redeem for
+    uint256 private amountVoltIn = 10_000e18;
+
+    /// @notice amount of dai to mint volt with
+    uint256 private amountDaiIn = 1_000_000e18;
+
+    uint256[] private amountsIn = [1_000_000e18, 1_000_000e6];
+
+    ICore private core = ICore(MainnetAddresses.CORE);
+    IERC20 private dai = IERC20(MainnetAddresses.DAI);
+    IERC20 private usdc = IERC20(MainnetAddresses.USDC);
+    IERC20 private volt = IVolt(MainnetAddresses.VOLT);
+
+    IERC20[] private tokensIn = [dai, usdc];
+
+    function _redeem(PriceBoundPSM psm, IERC20 underlying) private {
+        uint256 startingUserUnderlyingBalance = underlying.balanceOf(
+            address(this)
+        );
+        uint256 startingPSMUnderlyingBalance = underlying.balanceOf(
+            address(psm)
+        );
+        uint256 startingUserVoltBalance = volt.balanceOf(address(this));
+        uint256 startingPSMVoltBalance = volt.balanceOf(address(psm));
+
+        volt.approve(address(psm), amountVoltIn);
+
+        uint256 minAmountOut = psm.getRedeemAmountOut(amountVoltIn);
+        uint256 amountOut = psm.redeem(
+            address(this),
+            amountVoltIn,
+            minAmountOut
+        );
+
+        uint256 endingUserVOLTBalance = volt.balanceOf(address(this));
+        uint256 endingUserUnderlyingBalance = underlying.balanceOf(
+            address(this)
+        );
+        uint256 endingPSMUnderlyingBalance = underlying.balanceOf(address(psm));
+        uint256 endingPSMVoltBalance = volt.balanceOf(address(psm));
+
+        require(
+            endingPSMUnderlyingBalance ==
+                startingPSMUnderlyingBalance - amountOut,
+            "RedeemVerification: a0"
```

Comment:
We could have a clearer error message "RedeemVerification: PSM Underlying Mismatch" & similar for Volt mismatch and the same for the user variations for that

Reply:
agreed we can improve this